File: Skyline.hs
module Skyline where

-- TIPOS DE DATOS

type Edificio = (Int,Int,Int)
type Coordenada = (Int,Int)
type Skyline = [Coordenada]

-- FUNCIONES PRINCIPALES

-- resuelveSkyline
-- Toma como entrada una lista de edificios y devuelve un Skyline, una lista de coordenadas que forman la linea de horizonte
-- Para ello, divide la lista de edificios en dos partes (funcion divide) y hace una llamada recursiva con cada una de las partes
-- combinando las soluciones obtenidas de cada parte (funcion combina)
-- si el resultado de dividir la lista de edificios es un solo edificio, transforma el edificio a Skyline (funcion edificioAskyline)
-- seria el caso base de la recursividad de la funcion resuelveSkyline
resuelveSkyline :: [Edificio] -> Skyline
resuelveSkyline [] = []
resuelveSkyline [x] = edificioAskyline(x)
resuelveSkyline listaEdificios = combina (resuelveSkyline (fst (divide listaEdificios))) (resuelveSkyline (snd (divide listaEdificios))) 

-- edificioAskyline
-- Toma como entrada un edificio y devuelve su Skyline
edificioAskyline :: Edificio -> Skyline
edificioAskyline (x1,x2,h) = [(x1,h),(x2,0)]

-- divide
-- Toma como entrada una lista de edificios y devuelve una tupla formada por dos listas de edificios
-- Cada una de las 2 listas de edificios de salida es la mitad de la lista de edificios de entrada
divide :: [Edificio] -> ([Edificio],[Edificio])
divide [] = ([],[])
divide edificios = splitAt ((length edificios) `div` 2) edificios

-- combina
-- Toma como entrada 2 lineas de horizonte Skyline y devuelve la linea de hotrizonte Skyline como resultado de combinar
-- las 2 lineas de horizonte de entrada
-- Para ello hace uso de la funcion subcombina
combina :: Skyline -> Skyline -> Skyline
combina lineaHorizonte1 lineaHorizonte2 = subcombina lineaHorizonte1 lineaHorizonte2 0 0 0

-- FUNCIONES AUXILIARES

-- ***************************************************************************************
-- subcombina
-- Toma como parametros de entrada 2 lineas de horizonte Skyline, la altura del punto anterior de la linea de horizonte 1,
-- la altura del punto anterior de la linea de horizonte 2 y la altura del ultimo punto extraido de las lineas de horizonte.
-- Devuelve como resultado una linea de horizonte Skyline que es la combinacion de las 2 lineas de horizonte de entrada.
-- Para ello utiliza el proceso descrito en la seccion 2.1.1 del enunciado de la practica.
subcombina :: Skyline -> Skyline -> Int -> Int -> Int -> Skyline
subcombina [] [] _ _ _= []
subcombina [] lineaHorizonte2 _ _ _= lineaHorizonte2 
subcombina lineaHorizonte1 [] _ _ _= lineaHorizonte1
subcombina lineaHorizonte1 lineaHorizonte2 hAnterior1 hAnterior2 ultimaHextraida
     | obtenerX (primerElemento lineaHorizonte1) < obtenerX (primerElemento lineaHorizonte2) = 
        if (ultimaHextraida == (alturaMasGrande lineaHorizonte1 lineaHorizonte2 hAnterior2) ) then
            subcombina (tail lineaHorizonte1) lineaHorizonte2 
                       (obtenerH (primerElemento lineaHorizonte1)) hAnterior2 ultimaHextraida
        else
            [(puntoXmasPequeño lineaHorizonte1 lineaHorizonte2,
            alturaMasGrande lineaHorizonte1 lineaHorizonte2 hAnterior2)] ++ 
            subcombina (tail lineaHorizonte1) lineaHorizonte2 
                       (obtenerH (primerElemento lineaHorizonte1)) hAnterior2 (alturaMasGrande lineaHorizonte1 lineaHorizonte2 hAnterior2)
          
     | obtenerX (primerElemento lineaHorizonte1) > obtenerX (primerElemento lineaHorizonte2) = 
        if (ultimaHextraida == (alturaMasGrande lineaHorizonte1 lineaHorizonte2 hAnterior1) ) then
            subcombina lineaHorizonte1 (tail lineaHorizonte2) 
                       hAnterior1 (obtenerH (primerElemento lineaHorizonte2)) ultimaHextraida
        else
            [(puntoXmasPequeño lineaHorizonte1 lineaHorizonte2,
            alturaMasGrande lineaHorizonte1 lineaHorizonte2 hAnterior1)] ++ 
            subcombina lineaHorizonte1 (tail lineaHorizonte2) 
                       hAnterior1 (obtenerH (primerElemento lineaHorizonte2)) (alturaMasGrande lineaHorizonte1 lineaHorizonte2 hAnterior1)

     |otherwise =
        if (ultimaHextraida == (alturaMasGrande lineaHorizonte1 lineaHorizonte2 0) ) then
            subcombina (tail lineaHorizonte1) (tail lineaHorizonte2) 
                       (obtenerH (primerElemento lineaHorizonte1)) (obtenerH (primerElemento lineaHorizonte2)) ultimaHextraida
        else
            [(puntoXmasPequeño lineaHorizonte1 lineaHorizonte2,
            alturaMasGrande lineaHorizonte1 lineaHorizonte2 0)] ++ 
            subcombina (tail lineaHorizonte1) (tail lineaHorizonte2) 
                       (obtenerH (primerElemento lineaHorizonte1)) (obtenerH (primerElemento lineaHorizonte2)) (alturaMasGrande lineaHorizonte1 lineaHorizonte2 0)
-- ***************************************************************************************

-- ***************************************************************************************
-- puntoXmasPequeño
-- devuelve el la coordenada x del punto mas pequeño en la coordenada x
-- para ello, obtiene el primer elemento de cada linea de horizonte
-- obtiene el menor de los dos elementos anteriores en su coordenada x
-- y devuelve la coordenada x 
puntoXmasPequeño :: Skyline -> Skyline -> Int
puntoXmasPequeño lHorizonte1 lHorizonte2 = obtenerX (seleccionaElemento (primerElemento lHorizonte1) (primerElemento lHorizonte2))
-- ***************************************************************************************


-- ***************************************************************************************
-- selecciona el elemento con la coordenada x mas pequeña
-- si ambos tienen la misma coordenada, selecciona el elemento de mayor altura
seleccionaElemento :: Coordenada -> Coordenada -> Coordenada
seleccionaElemento ele1 ele2 
     | obtenerX ele1 < obtenerX ele2 = ele1
     | obtenerX ele2 < obtenerX ele1 = ele2
     | otherwise = elementoAlturaMasGrande ele1 ele2
-- ***************************************************************************************


-- ***************************************************************************************
-- elementoAlturaMasGrande
-- selecciona el elemento con la altura mas grande
elementoAlturaMasGrande :: Coordenada -> Coordenada -> Coordenada
elementoAlturaMasGrande ele1 ele2 
     | obtenerH ele1 > obtenerH ele2 = ele1
     | obtenerH ele2 > obtenerH ele1 = ele2
     | otherwise = ele1
-- ***************************************************************************************

-- ***************************************************************************************
-- alturaMasGrande
-- devuelve la altura mas grande de dos alturas dadas
-- para ello, obtiene el primer elemento de cada lista de horizonte
-- obtiene el menor de los dos elementos anteriores en su coordenada x
-- compara la altura del elemento obtenido con la altura del ultimo elemento consumido
-- y devuelve la mayor de las alturas
alturaMasGrande :: Skyline -> Skyline -> Int -> Int
alturaMasGrande lHorizonte1 lHorizonte2 hAnterior = maximo (obtenerH (seleccionaElemento (primerElemento lHorizonte1) (primerElemento lHorizonte2))) hAnterior
-- ***************************************************************************************

-- ***************************************************************************************
-- Toma como entrada 2 valores enteros y devuelve el mayor de los 2
maximo :: Int -> Int -> Int
maximo hActual hAnterior 
     | hActual > hAnterior = hActual
     | hAnterior > hActual = hAnterior
     | otherwise = hActual
-- ***************************************************************************************

-- ***************************************************************************************
-- obtenerX devuelve la X de una coordenada
obtenerX (x,_) = x
-- obtenerH devuelve la altura de una coordenada
obtenerH (_,h) = h
-- ***************************************************************************************


-- ***************************************************************************************
-- primerElemento
-- devuelve el primer elemento de una linea de horizonte
-- devuelve la primera coordenada
primerElemento :: Skyline -> Coordenada
primerElemento x = head x
-- ***************************************************************************************

-- CUESTIONES SOBRE LA PRACTICA
-- 1a. FUNCION DIBUJA SKYLINE

-- dibujaSkyline 
-- Toma como entrada un Skiline y lo transforma en un string de * y caracteres en blanco
-- el * representa la presencia de edificio y el caracter blanco la ausencia de este
-- el string termina con una linea de guiones que representan el suelo
dibujaSkyline :: Skyline -> String
dibujaSkyline listaSkyline =
     "" ++ (subDibujaSkyline (listaDeAlturas listaSkyline 0 0) (alturaMaximaSkyline listaSkyline 0))


-- subDibujaSkyline
-- Toma como entrada un Skyline transformado en una lista de alturas y la linea a dibujar
-- y devuelve un string con la linea a mostrar en pantalla
subDibujaSkyline :: Skyline -> Int -> String
subDibujaSkyline listaH 0 = dibujaGuiones listaH
subDibujaSkyline listaH altura =
     (dibujaLinea listaH altura) ++ (subDibujaSkyline listaH (altura - 1))

-- listaDeAlturas
-- Transforma un Skyline en una lista (tambien Skyline) con las alturas
-- de todas las coordenadas x del Skyline
-- para ello, para ello, comprueba si cada coordenada x es un punto del Skiline
-- si lo es, añade el punto del skyline a la lista de alturas
-- si no lo es, añade a la lista de alturas la coordena x que se esta comprobando junto con la altura del ultimo
-- punto encontrado del skyline
listaDeAlturas :: Skyline -> Int -> Int -> Skyline
listaDeAlturas [] _ _= []
--listaDeAlturas [x] _= [x]
listaDeAlturas listaH coordenadaX alturaH= 
     if (obtenerX (primerElemento listaH) == coordenadaX) then
          [primerElemento listaH] ++ listaDeAlturas (tail listaH) (coordenadaX + 1) (obtenerH(primerElemento listaH)) 
     else
          [(coordenadaX,alturaH)] ++ listaDeAlturas listaH (coordenadaX + 1) alturaH


-- alturaMaximaSkyline
-- devuelve la altura maxima del skyline
-- para ello, almacena la altura maxima encontrada y la compara con la altura del punto x del skyline que se esta examinando
-- si el punto tiene una altura mayor, actualiza la altura almacenada con la altura del punto examinado
-- y continua con el siguiente punto del skyline
alturaMaximaSkyline :: Skyline -> Int -> Int
alturaMaximaSkyline [] alturaMax = alturaMax
alturaMaximaSkyline listaH alturaMax =
     if (alturaMax < obtenerH(primerElemento listaH)) then
          alturaMaximaSkyline (tail listaH) (obtenerH(primerElemento listaH))
     else
          alturaMaximaSkyline (tail listaH) alturaMax

-- dibujaLinea
-- Toma como entrada un Skyline y la altura de este a dibujar
-- y devuelve un string con la linea a mostrar en pantalla formado por * para las zonas donde hay edificio
-- y espacios en blanco para las zonas donde no hay edificios
-- muestra una linea de guiones para la altura 0, representando el suelo sobre el que se alza el edificio
dibujaLinea :: Skyline -> Int -> String
dibujaLinea listaSkyline 0 = dibujaGuiones listaSkyline
dibujaLinea [] _ = "\n"
dibujaLinea listaSkyline altura =
     if ( (obtenerH(primerElemento listaSkyline)) >= altura) then
          "*" ++ (dibujaLinea (tail listaSkyline) altura)
     else
          " " ++ (dibujaLinea (tail listaSkyline) altura)

-- dibujaGuiones
-- dibuja una linea de guiones representando el suelo sobre el que se alza el edificio
dibujaGuiones :: Skyline -> String
dibujaGuiones [] = ""
dibujaGuiones listaSkyline = "-" ++ (dibujaGuiones (tail listaSkyline))

-- muestra en pantalla un string
write :: String -> IO()
write [] = putChar '\n'
write (x:xs) = putChar x >> write xs

File: Skyline.pl
%Predicados a implementar

%------------------------------------------
% Implementa el predicado principal de la practica.
% Toma una lista de edificios y devuelve su linea de horizinte.
resuelveSkyline([],[]).
resuelveSkyline([Ed],LineaH) :- edificioAskyLine(Ed,LineaH).
resuelveSkyline(ListaEd,LineaH) :- divide(ListaEd,ListaEd1,ListaEd2),
				   resuelveSkyline(ListaEd1,LineaH1),
				   resuelveSkyline(ListaEd2,LineaH2),
				   combina(LineaH1,LineaH2,LineaH).
%------------------------------------------

%------------------------------------------
% Transforma un edificio a su linea de horizonte.
%  Toma un edificio de la forma ed(X1,X2,H) y devuelve la linea de
%  horizonte del edificio en la forma [c(X1,H),c(X2,0)]
edificioAskyLine(ed(X1,X2,H),[c(X1,H),c(X2,0)]).
%------------------------------------------

%------------------------------------------
% Divide una lista de edificios en dos listas de edificios
divide([],[],[]).
divide(ListaEd,ListaEd1,ListaEd2) :- append(ListaEd1,ListaEd2,ListaEd),
				     length(ListaEd,Long0),
				     length(ListaEd1,Long1),
				     Mitad is Long0//2,
				     Long1==Mitad.
%------------------------------------------

%------------------------------------------
% Combina dos lineas de horizonte en una sola linea de horizonte
combina(LineaH1,LineaH2,LineaH) :- subCombina(LineaH1,LineaH2,0,0,0,LH),
				   limpiarLH(LH,LineaH).
%------------------------------------------

% Predicados Auxiliares

%------------------------------------------
% Combina dos lineas de horizonte en una sola linea de horizonte.
% Toma como parametros las dos lineas de horizonte a combinar, la altura
% del primer punto de cada linea de horizonte tratados en el ciclo
% anterior y la ultima altura que se ha extraido de una de las lineas de
% horizonte. La combinacion de las dos lineas de horizonte se realiza
% segun el proceso descrito den la seccion 2.1.1 del enunciado de la
% practica. En caso de que en un ciclo no sea necesario añadir ningun
% punto nuevo a la linea de horizonte resultante, se añadira un punto
% nulo de la forma c(0,0), de esta forma, se evita que prolog genere una
% evaluacion false y aborte el programa. Posteriormente, se eliminaran
% de la linea de horizonte resultante estos puntos nulos.
subCombina([],L2,_,_,_,L2).
subCombina(L1,[],_,_,_,L1).
subCombina([],[],_,_,_,[]).
subCombina(L1,L2,Hanterior1,Hanterior2,UltimaHextraida,R2) :-
        L1=[Cab1|Col1], L2=[Cab2|Col2],
	obtenerX(Cab1,CX1),
	obtenerX(Cab2,CX2),
	(
	    (   CX1<CX2 ->
	             alturaMasGrande(Cab1,c(0,0),Hanterior2,HmasGrande),
		     (
			 (   UltimaHextraida==HmasGrande ->
			          obtenerH(Cab1,H1),
			          subCombina(Col1,L2,H1,Hanterior2,UltimaHextraida,R),
			          añadir(c(0,0),R,R2)
			  );
			 (   UltimaHextraida\=HmasGrande ->
			          obtenerH(Cab1,H1),
			          subCombina(Col1,L2,H1,Hanterior2,HmasGrande,R),
			          obtenerCoordenadaXmasPequeña(Cab1,Cab2,CoordenadaX),
			          crearPunto(CoordenadaX,HmasGrande,Punto),
			          añadir(Punto,R,R2)
			  )
		     )
	    );
	    (   CX1>CX2 ->
	             alturaMasGrande(c(0,0),Cab2,Hanterior1,HmasGrande),
		     (
			 (   UltimaHextraida==HmasGrande ->
			          obtenerH(Cab2,H2),
			          subCombina(L1,Col2,Hanterior1,H2,UltimaHextraida,R),
			          añadir(c(0,0),R,R2)
			  );
			 (   UltimaHextraida\=HmasGrande ->
			          obtenerH(Cab2,H2),
			          subCombina(L1,Col2,Hanterior1,H2,HmasGrande,R),
			          obtenerCoordenadaXmasPequeña(Cab1,Cab2,CoordenadaX),
			          crearPunto(CoordenadaX,HmasGrande,Punto),
			          añadir(Punto,R,R2)
			  )
		     )
	    );
	    (	CX1==CX2 ->
	             alturaMasGrande(Cab1,Cab2,0,HmasGrande),
		     (
			 (   UltimaHextraida==HmasGrande ->
			          obtenerH(Cab1,H1),
			          obtenerH(Cab2,H2),
			          subCombina(Col1,Col2,H1,H2,UltimaHextraida,R),
			          añadir(c(0,0),R,R2)
			  );
			 (   UltimaHextraida\=HmasGrande ->
			          obtenerH(Cab1,H1),
			          obtenerH(Cab2,H2),
			          subCombina(Col1,Col2,H1,H2,HmasGrande,R),
				  obtenerCoordenadaXmasPequeña(Cab1,Cab2,CoordenadaX),
			          crearPunto(CoordenadaX,HmasGrande,Punto),
			          añadir(Punto,R,R2)
			  )
		     )
	    )
	).
%------------------------------------------

%------------------------------------------
% Elimina de la lina de horizonte los puntos nulos de la forma c(0,0)
% introducidos por el predicado subcombina.
limpiarLH([],[]).
limpiarLH([Cab|Col],LH) :-    (   Cab\=c(0,0) -> limpiarLH(Col,R),
			                         añadir(Cab,R,LH)
			      );
                              (   Cab==c(0,0) -> limpiarLH(Col,R),
			                         LH=R
			      ).
%------------------------------------------

%------------------------------------------
% Crea un nuevo punto, coordenada, de la forma c(X,H), siendo X la
% coordenada del eje X y H la altura.
crearPunto(Cx,Ch,c(Cx,Ch)).
%------------------------------------------

%------------------------------------------
% Dados 2 puntos en la forma c(X,H), devuelve la coordenada X mas
% pequeña de los dos puntos.
obtenerCoordenadaXmasPequeña(c(X1,_),c(X2,_),X) :- (X1>X2 -> X=X2);
                                                   (X1<X2 -> X=X1);
						   (X1==X2 -> X=X1).
%------------------------------------------

%------------------------------------------
% Dados dos puntos en la forma c(X,H) y una altura H, devuelve la altura
% mas grande de los dos puntos y la altura H.
alturaMasGrande(C1,C2,H,Hmax) :- C1=c(_,H1), C2=c(_,H2),
				 valorMayor(H1,H2,Haux),
				 valorMayor(Haux,H,Hmax).
%------------------------------------------

%------------------------------------------
% Dados 2 valores X Y, devuelve el mayor de los dos
valorMayor(X,Y,R) :- (X>Y -> R=X);
                     (X<Y -> R=Y);
		     (X==Y -> R=X).
%------------------------------------------

%------------------------------------------
% Dado un punto en la forma c(X,H), devuelve el valor de la coordenada X
% del punto
obtenerX(E,CX) :- E=c(CX,_).
%------------------------------------------

%------------------------------------------
%Dado un punto en la forma c(X,H), devuelve la altura H del punto
obtenerH(E,H) :- E=c(_,H).
%------------------------------------------

%------------------------------------------
% Dado un elemento E y una lista L1, añade el elemento E al principio de
% la lista L1
añadir(E,[],[E]).
añadir(E,L1,[E|L1]).
%------------------------------------------

% Cuestiones sobre la practica. Parte Optativa.
% Punto 1a. Dibuja Skyline.

%------------------------------------------
% Muestra en pantalla un dibujo de la la linea de horizonte dada
% Para ello transforma la linea de horizonte en una lista de alturas,
% segun lo especificado en el punto 3 del enunciado de la practica.
% Obtiene la mayor altura de la linea de horizonte y la dibuja en
% pantalla
dibujaSkyline(LineaH) :- obtenerListaAlturas(LineaH,0,c(0,0),LA),
	                 obtenerAlturaMaxima(LineaH,AlturaSkyline),
                         dibuja(LA,AlturaSkyline).
%------------------------------------------

%------------------------------------------
% Dibuja en pantalla la linea de horizonte dada, linea por linea,
% comenzando por la linea de mayor altura hasta llegar a la linea de
% altura 0 en la que dibujara el suelo
dibuja(ListaH,0) :- dibujaSuelo(ListaH).
dibuja(ListaH,Altura) :- Alt is Altura-1,
	                 dibujaLinea(ListaH,Altura),
			 dibuja(ListaH,Alt).
%------------------------------------------

%------------------------------------------
% Dibuja la linea de la altura especificada por Altura
dibujaLinea([],_) :- write_ln(" ").
dibujaLinea(ListaH,Altura) :- ListaH = [Cab|Col],
			      obtenerH(Cab,H),
			      dibujaPunto(H,Altura),
			      dibujaLinea(Col,Altura).
%------------------------------------------

%------------------------------------------
% Dibuja el suelo del skyline en forma de - en la altura 0
dibujaSuelo([]) :- write_ln(" ").
dibujaSuelo(ListaH) :- ListaH = [_|Col],
		       write("-"),
		       dibujaSuelo(Col).
%------------------------------------------

%------------------------------------------
% Dibuja un punto en pantalla, muestra un * si hay edificio y un espacio
% blanco si no lo hay
dibujaPunto(P,Alt) :- (   P > Alt -> write("*"));
		      (	  P == Alt -> write("*"));
		      (	  P < Alt -> write(" ")).
%------------------------------------------

%------------------------------------------
% Transforma la linea de horizonte en una lista de alturas, segun lo
% especificado en el apartado 3 del eneunciado de la practica.
% En la lista de alturas, todas las coordenadas del eje X tienen un
% punto. En la linea de horizonte solo tiene un punto aquellas
% coordenadas del eje X que coincidan con un cambio de altura.
% En la linea de alturas, las coordenadas el eje X que no coincidan con
% un cambio de altura, tendran la misma altura que la ultima coordenada
% X que tenga un cambio de altura
obtenerListaAlturas([],_,_,[]).
obtenerListaAlturas(LineaH,CoX,PuntoAnterior,R2) :-
	   LineaH = [Cab|Col],
	   obtenerX(Cab,CX),
	   X is CoX+1,
	   (
	       (   CoX==CX ->
	                      obtenerListaAlturas(Col,X,Cab,R),
	                      añadir(Cab,R,R2)
	        );
	       (   CoX\=CX ->
			      obtenerListaAlturas(LineaH,X,PuntoAnterior,R),
		              obtenerH(PuntoAnterior,Hanterior),
		              crearPunto(CoX,Hanterior,Punto),
	                      añadir(Punto,R,R2)
	        )
	   ).
%------------------------------------------

%------------------------------------------
% Obtiene la altura mayor de una linea de horizonte.
obtenerAlturaMaxima([],0).
obtenerAlturaMaxima(LineaH,Hmax) :-
	LineaH = [Cab|Col],
        obtenerH(Cab,H),
	obtenerAlturaMaxima(Col,Haux),
	valorMayor(H,Haux,Hmax).
%------------------------------------------

























































